cpp/rss_subset_aggregator.cc

```diff
@@ -0,0 +1,315 @@
+    XmlWriter xml_writer(FileUtil::OpenOutputFileOrDie("/dev/stdout").release(),
+                         XmlWriter::WriteTheXmlDeclaration, DEFAULT_XML_INDENT_AMOUNT);
+    WriteRSSFeedXMLOutput(subsystem_type, harvested_items, &xml_writer);
+}
+
+
+bool ProcessFeeds(const std::string &user_id, const std::string &rss_feed_last_notification, const std::string &email_sender,
+                  const std::string &user_email, const std::string &user_address, const std::string &language,
+                  const bool send_email, const std::string &subsystem_type,
+                  DbConnection * const db_connection)
+{
+    db_connection->queryOrDie("SELECT rss_feeds_id FROM tuefind_rss_subscriptions WHERE user_id=" + user_id);
+    auto rss_subscriptions_result_set(db_connection->getLastResultSet());
+    std::vector<std::string> feed_ids;
+    while (const auto row = rss_subscriptions_result_set.getNextRow())\
+        feed_ids.emplace_back(row["rss_feeds_id"]);
+    if (feed_ids.empty())
+        return false;
+
+    std::vector<HarvestedRSSItem> harvested_items;
+    std::string max_insertion_time;
+    for (const auto &feed_id : feed_ids) {
+        db_connection->queryOrDie("SELECT feed_name,feed_url FROM tuefind_rss_feeds WHERE id=" + feed_id);
+        auto feed_result_set(db_connection->getLastResultSet());
+        const auto feed_row(feed_result_set.getNextRow());
+        const auto feed_name(feed_row["feed_name"]);
+        const auto feed_url(feed_row["feed_url"]);
+        feed_result_set.~DbResultSet();
+
+        db_connection->queryOrDie("SELECT item_title,item_description,item_url,item_id,pub_date,insertion_time FROM "
+                                  "tuefind_rss_items WHERE rss_feeds_id=" + feed_id + " AND insertion_time > '"
+                                  + rss_feed_last_notification + "'");
+        auto items_result_set(db_connection->getLastResultSet());
+        while (const auto item_row = items_result_set.getNextRow()) {
+            harvested_items.emplace_back(SyndicationFormat::Item(item_row["item_title"], item_row["item_description"],
+                                                                 item_row["item_url"], item_row["item_id"],
+                                                                 SqlUtil::DatetimeToTimeT(item_row["pub_date"])),
+                                         feed_name, feed_url);
+            const auto insertion_time(item_row["insertion_time"]);
+            if (insertion_time > max_insertion_time)
+                max_insertion_time = insertion_time;
+        }
+    }
+    if (harvested_items.empty())
+        return false;
+
+    if (send_email) {
+        if (not SendEmail(subsystem_type, email_sender, user_email, user_address, language, harvested_items))
+            return true;
+        db_connection->queryOrDie("INSERT INTO user SET tuefind_rss_feed_last_notification='" + max_insertion_time
```

Comment:
This should be an UPDATE statement and no INSERT statement

Reply:
Done